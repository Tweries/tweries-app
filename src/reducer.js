import initialState from './initialState';
import makeItem from './makeItem';

function reducer(state = initialState, action) {
  switch (action.type) {
    case 'ADD_ITEM': {
      return {
        ...state,
        items: [
          ...state.items.map(
            item =>
              (item = makeItem(
                state.hashtags,
                item.id,
                state.items.length + 1,
                item.position,
                item.source
              ))
          ),
          makeItem(
            state.hashtags,
            null,
            state.items.length + 1,
            state.items.length + 1
          )
        ]
      };
    }
    case 'CHANGE_SOURCE': {
      return {
        ...state,
        items: state.items.map(item => {
          if (item.id === action.value.id) {
            item = makeItem(
              state.hashtags,
              item.id,
              state.items.length,
              item.position,
              action.value.source
            );
          }
          return item;
        })
      };
    }
    case 'REMOVE_ITEM': {
      const copy = [...state.items];
      copy.pop();
      return {
        ...state,
        items: copy.map(
          item =>
            (item = makeItem(
              state.hashtags,
              item.id,
              copy.length,
              item.position,
              item.source
            ))
        )
      };
    }
    case 'UPDATE_HASHTAGS': {
      return {
        ...state,
        hashtags: action.value,
        items: state.items.map(
          item =>
            (item = makeItem(
              action.value,
              item.id,
              state.items.length,
              item.position,
              item.source
            ))
        )
      };
    }
    default:
      return state;
  }
}

export default reducer;
