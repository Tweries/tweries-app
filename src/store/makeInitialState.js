import { NEWLINE } from '../constants';
import makeTweetstorm from './makeTweetstorm';

function makeInitialState({ feature }) {
  return {
    healthy: false,
    items: makeTweetstorm(feature)({
      linefeed: NEWLINE,
      source: ''
    }),
    linefeed: NEWLINE,
    notification: null,
    source: '',
    userId: ''
  };
}

export default makeInitialState;
