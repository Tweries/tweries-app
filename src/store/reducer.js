import makeInitialState from './makeInitialState';
import makeTweetstorm from './makeTweetstorm';

export const types = {
  CHANGE_HASHTAGS: 'CHANGE_HASHTAGS',
  CHANGE_LINEFEED: 'CHANGE_LINEFEED',
  CHANGE_SOURCE: 'CHANGE_SOURCE',
  RESET_TWEETSTORM: 'RESET_TWEETSTORM',
  SET_HEALTHY: 'SET_HEALTHY',
  SET_USER_ID: 'SET_USER_ID'
};

function reducer(state = makeInitialState({}), action) {
  switch (action.type) {
    case types.CHANGE_HASHTAGS: {
      return {
        ...state,
        hashtags: action.value,
        items: makeTweetstorm({
          feature: state.feature,
          hashtags: action.value,
          linefeed: state.linefeed,
          source: state.source
        })
      };
    }
    case types.CHANGE_LINEFEED: {
      return {
        ...state,
        linefeed: action.value,
        items: makeTweetstorm({
          feature: state.feature,
          hashtags: state.hashtags,
          linefeed: action.value,
          source: state.source
        })
      };
    }
    case types.CHANGE_SOURCE: {
      return {
        ...state,
        items: makeTweetstorm({
          feature: state.feature,
          hashtags: state.hashtags,
          linefeed: state.linefeed,
          source: action.value
        }),
        source: action.value
      };
    }
    case types.RESET_TWEETSTORM: {
      return {
        ...makeInitialState({}),
        healthy: state.healthy,
        userId: state.userId
      };
    }
    case types.SET_HEALTHY: {
      return {
        ...state,
        healthy: action.value
      };
    }
    case types.SET_USER_ID: {
      return {
        ...state,
        userId: action.value
      };
    }
    default:
      return state;
  }
}

export default reducer;
