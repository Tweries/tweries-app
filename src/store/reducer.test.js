import reducer, { types } from './reducer';
import makeInitialState from './makeInitialState';

const USER_ID = 'twitter|1183836409850814464';

describe('reducer', () => {
  const scenarios = [
    {
      description: types.CHANGE_HASHTAGS,
      props: {
        action: {
          type: types.CHANGE_HASHTAGS,
          value: '#FOO'
        },
        state: { ...makeInitialState({ source: 'bar' }) }
      }
    },
    {
      description: types.CHANGE_LINEFEED,
      props: {
        action: {
          type: types.CHANGE_LINEFEED,
          value: '||'
        },
        state: {
          ...makeInitialState({
            source: 'Can you can a can [..] as a canner || can can a can?'
          })
        }
      }
    },
    {
      description: types.CHANGE_SOURCE,
      props: {
        action: {
          type: types.CHANGE_SOURCE,
          value: 'bar'
        },
        state: undefined
      }
    },
    {
      description: types.RESET_TWEETSTORM,
      props: {
        action: {
          type: types.RESET_TWEETSTORM
        },
        state: {
          ...makeInitialState({ hashtags: '#FOO', source: 'bar' }),
          healthy: true,
          userId: USER_ID
        }
      }
    },
    {
      description: types.SET_HEALTHY,
      props: {
        action: {
          type: types.SET_HEALTHY,
          value: true
        },
        state: undefined
      }
    },
    {
      description: types.SET_USER_ID,
      props: {
        action: {
          type: types.SET_USER_ID,
          value: USER_ID
        },
        state: undefined
      }
    },
    {
      description: 'default',
      props: {
        action: { type: 'UNKNOWN', value: undefined },
        state: undefined
      }
    }
  ];

  scenarios.forEach(({ description, props: { action, state } }) => {
    it(description, () => {
      expect(reducer(state, action)).toMatchSnapshot();
    });
  });
});
