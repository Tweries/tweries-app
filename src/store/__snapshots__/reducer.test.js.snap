// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer CHANGE_HASHTAGS 1`] = `
Object {
  "feature": Object {
    "active": [Function],
  },
  "hashtags": "#FOO",
  "healthy": false,
  "items": Array [
    Object {
      "length": 9,
      "tweet": "bar #FOO ",
    },
  ],
  "linefeed": ";;",
  "notification": null,
  "source": "bar",
  "userId": null,
}
`;

exports[`reducer CHANGE_LINEFEED 1`] = `
Object {
  "feature": Object {
    "active": [Function],
  },
  "hashtags": "",
  "healthy": false,
  "items": Array [
    Object {
      "length": 38,
      "tweet": "Can you can a can [..] as a canner 1/2",
    },
    Object {
      "length": 18,
      "tweet": "can can a can? 2/2",
    },
  ],
  "linefeed": "||",
  "notification": null,
  "source": "Can you can a can [..] as a canner || can can a can?",
  "userId": null,
}
`;

exports[`reducer CHANGE_SOURCE 1`] = `
Object {
  "feature": Object {
    "active": [Function],
  },
  "hashtags": "",
  "healthy": false,
  "items": Array [
    Object {
      "length": 4,
      "tweet": "bar ",
    },
  ],
  "linefeed": ";;",
  "notification": null,
  "source": "bar",
  "userId": null,
}
`;

exports[`reducer DISMISS_TOAST 1`] = `
Object {
  "feature": Object {
    "active": [Function],
  },
  "hashtags": "",
  "healthy": false,
  "items": Array [],
  "linefeed": ";;",
  "notification": null,
  "source": "",
  "userId": null,
}
`;

exports[`reducer RESET_TWEETSTORM 1`] = `
Object {
  "feature": Object {
    "active": [Function],
  },
  "hashtags": "",
  "healthy": true,
  "items": Array [],
  "linefeed": ";;",
  "notification": Object {
    "message": "Oh Noes!",
    "type": "danger",
  },
  "source": "",
  "userId": "twitter|1183836409850814464",
}
`;

exports[`reducer SET_HEALTHY 1`] = `
Object {
  "feature": Object {
    "active": [Function],
  },
  "hashtags": "",
  "healthy": true,
  "items": Array [],
  "linefeed": ";;",
  "notification": null,
  "source": "",
  "userId": null,
}
`;

exports[`reducer SET_USER_ID 1`] = `
Object {
  "feature": Object {
    "active": [Function],
  },
  "hashtags": "",
  "healthy": false,
  "items": Array [],
  "linefeed": ";;",
  "notification": null,
  "source": "",
  "userId": "twitter|1183836409850814464",
}
`;

exports[`reducer default 1`] = `
Object {
  "feature": undefined,
  "hashtags": "",
  "healthy": false,
  "items": Array [],
  "linefeed": ";;",
  "notification": null,
  "source": "",
  "userId": null,
}
`;
